src/gflownet/data/data_source.py

```diff
@@ -0,0 +1,320 @@
+                cond_info = self.task.sample_conditional_information(num_samples, t)
+                objs, props = map(list, zip(*[data[i] for i in idcs])) if len(idcs) else ([], [])
+                trajs = self.algo.create_training_data_from_graphs(objs, backwards_model, cond_info["encoding"], p)
+                self.set_traj_cond_info(trajs, cond_info)  # Attach the cond info to the trajs
+                self.set_traj_props(trajs, props)
+                self.compute_log_rewards(trajs)
+                yield trajs, {}
+
+        self.iterators.append(iterator)
+        return self
+
+    def do_conditionals_dataset_in_order(self, data, num_samples, model):
+        def iterator():
+            for idcs in self.iterate_indices(len(data), num_samples):
+                t = self.current_iter
+                p = self.algo.get_random_action_prob(t)
+                # TODO: when we refactor cond_info, data[i] will probably be a dict? (or CondInfo objects)
+                # I'm also not a fan of encode_conditional_information, it assumes lots of things about what's passed to
+                # it and the state of the program (e.g. validation mode)
+                cond_info = self.task.encode_conditional_information(torch.stack([data[i] for i in idcs]))
+                trajs = self.algo.create_training_data_from_own_samples(model, len(idcs), cond_info["encoding"], p)
+                self.set_traj_cond_info(trajs, cond_info)  # Attach the cond info to the trajs
+                self.compute_properties(trajs, mark_as_online=True)
+                self.compute_log_rewards(trajs)
+                self.send_to_replay(trajs)  # This is a no-op if there is no replay buffer
+                # If we're using a dataset of preferences, the user/hooks may want to know the id of the preference
+                for i, j in zip(trajs, idcs):
+                    i["data_idx"] = j
+                batch_info = self.call_sampling_hooks(trajs)
+                yield trajs, batch_info
+
+        self.iterators.append(iterator)
+        return self
+
+    def do_sample_dataset(self, data, num_samples, backwards_model):
+        def iterator():
+            while self.active:
+                idcs = self.sample_idcs(len(data), num_samples)
+                t = self.current_iter
+                p = self.algo.get_random_action_prob(t)
+                cond_info = self.task.sample_conditional_information(num_samples, t)
+                objs, props = map(list, zip(*[data[i] for i in idcs])) if len(idcs) else ([], [])
+                trajs = self.algo.create_training_data_from_graphs(objs, backwards_model, cond_info["encoding"], p)
+                self.set_traj_cond_info(trajs, cond_info)  # Attach the cond info to the trajs
+                self.set_traj_props(trajs, props)
+                self.compute_log_rewards(trajs)
+                yield trajs, {}
+
+        self.iterators.append(iterator)
+        return self
```

Comment:
Less clear to me when this one is used (only for QM9 I believe?). Is it meant to compute the traj_likelihood of the model for these fixed examples?

Reply:
No this just creates the trajectories (backwards from a starting x that comes from a dataset) using either a uniform P_B, or `backwards_model` if provided. It's up to the Algo to compute likelihoods and such.